### PR TITLE
Address Classificator CRUD functional added (BackEnd only)

### PR DESCRIPTION
**Pull request by YuriPushchalo**
Create, Read, Update, Delete functions for Address Classificator realised.
This functional realized only on BackEnd.
FrontEnd is coming soon.

### CODE REVIEW

File: persistence/src/main/java/com/softserve/osbb/repository/OsbbRepository.java [hunk 1]
```diff
@@ -17,4 +17,9 @@
     List<Osbb> findByAvailable(Boolean available);
 
     List<Osbb> findByNameContainingIgnoreCase(String name);
+
+    List<Osbb> findByStreetId(Integer streetId);
+
+    List<Osbb> findByDistrictId(Integer districtId);
+
 }
```
Comment (KostyantynPanchenko) [inline persistence/src/main/java/com/softserve/osbb/repository/OsbbRepository.java]: Add new line plz

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java [hunk 2]
```diff
@@ -81,4 +85,132 @@ public City getCityById(Integer id) {
...
+    @Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+    @Override
+    public Region addRegion(Region region) {
+        if(region == null) {
+            throw new IllegalArgumentException("Region is null. Try to set correct data.");
+        }
+        return regionRepository.saveAndFlush(region);
+    }
+
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java:90]: Check indentation

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java [hunk 3]
```diff
@@ -81,4 +85,132 @@ public City getCityById(Integer id) {
 	public Region getRegionById(Integer id) {
 		return regionRepository.findById(id);
 	}
+ 
+    @Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+    @Override
+    public Region addRegion(Region region) {
+        if(region == null) {
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java]: Add space after if before (

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java [hunk 4]
```diff
@@ -81,4 +85,132 @@ public City getCityById(Integer id) {
...
+            throw new IllegalArgumentException("Region is null. Try to set correct data.");
+        }
+        return regionRepository.saveAndFlush(region);
+    }
+
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+    @Override
+    public Region updateRegion(Region region) {
+        if(regionRepository.exists(region.getId())) {
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java]: if (

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java [hunk 5]
```diff
@@ -81,4 +85,132 @@ public City getCityById(Integer id) {
...
+    	} else {
+            throw new IllegalArgumentException("Cities of region with id=" + id
+            + " exist. First they should be deleted.");
+    	}
+    }
+
+    @Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+    @Override
+    public City addCity(City city) {
+        if(city == null) {
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java]: if (

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java [hunk 6]
```diff
@@ -81,4 +85,132 @@ public City getCityById(Integer id) {
...
+            throw new IllegalArgumentException("City is null. Try to set correct data.");
+        }
+        return cityRepository.saveAndFlush(city);
+    }
+
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java:99]: Check indentation

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java [hunk 7]
```diff
@@ -81,4 +85,132 @@ public City getCityById(Integer id) {
...
+        if(district == null) {
+            throw new IllegalArgumentException("District is null. Try to set correct data.");
+        }
+        return districtRepository.saveAndFlush(district);
+    }
+
+	@Transactional(readOnly = false, propagation = Propagation.REQUIRED)
+    @Override
+    public District updateDistrict(District district) {
+        if(districtRepository.exists(district.getId())) {
```
Comment (KostyantynPanchenko) [inline service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java]: Space after if

File: web/src/main/java/com/softserve/osbb/controller/AddressController.java [hunk 8]
```diff
@@ -102,4 +105,91 @@
         return new ResponseEntity<>(district, status);
     }
 
+    @ResponseStatus(HttpStatus.OK)
```
Comment (KostyantynPanchenko) [inline web/src/main/java/com/softserve/osbb/controller/AddressController.java]: Why every method return HttpStatus.OK?
I don't see exception handler methods or other mechanism to handle negative situation

**Review by KostyantynPanchenko — CHANGES_REQUESTED** [review]
Please make changes or provide feedback

File: persistence/src/main/java/com/softserve/osbb/repository/OsbbRepository.java (hunk 1 above)
Comment (YuriPushchalo) [reply persistence/src/main/java/com/softserve/osbb/repository/OsbbRepository.java]: Done

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java (hunk 2 above)
Comment (YuriPushchalo) [reply service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java:90]: Done

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java (hunk 3 above)
Comment (YuriPushchalo) [reply service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java]: Done

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java (hunk 4 above)
Comment (YuriPushchalo) [reply service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java]: Done

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java (hunk 5 above)
Comment (YuriPushchalo) [reply service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java]: Done

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java (hunk 6 above)
Comment (YuriPushchalo) [reply service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java:99]: Done

File: service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java (hunk 7 above)
Comment (YuriPushchalo) [reply service/src/main/java/com/softserve/osbb/service/impl/AddressServiceImpl.java]: Done

File: web/src/main/java/com/softserve/osbb/controller/AddressController.java (hunk 8 above)
Comment (YuriPushchalo) [reply web/src/main/java/com/softserve/osbb/controller/AddressController.java]: ErrorHandler added

**Review by KostyantynPanchenko — APPROVED** [review]

**Review by aprudyus — APPROVED** [review]

**Review by thrytsko-amplify — APPROVED** [review]

**Pull request merged**